[API][IR] First stab at the backend translation

ggevay opened the pull request:
I tried to split this into digestable commits. The main part is in the last commit. This adds `TranslateToDataflows` to `Backend`, which
- Changes bag sources to the corresponding backend sources
- When a bag is used in a higher-order context then it inserts a fetch call to convert it to a local bag

This is enough to compile everything in `BaseCodegenTest` to `SparkRDD` and `FlinkDataSet`. (Except a few things like the statefuls, see comments there.) (Btw. the test failure that involved the structural types was caused by a case class being instantiated with `new` in the test. When I changed it to the usual way (`apply`), then the problem went away.)

What is missing at the moment:
- I'm using `desugar` instead of `combination` (because for `combination`, the `join` and `cross` combinators will need to be translated separately)
- Insert some `.cache` calls (we will need this e.g. for the examples that have control flow)
- `parallelize` macro

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala:
```diff
@@ -313,6 +318,34 @@ trait DataBag[A] extends Serializable {
+  }
+
+  override def hashCode(): Int =
+    scala.util.hashing.MurmurHash3.unorderedHash(fetch())
```
Maybe do it with a `fold`, see https://gist.github.com/aalexandrov/089f0fd4d3de13d37536f035dfe14c19

ggevay replied:
I would also leave it like this for the moment. But I'll add a TODO.

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+
+      val isHighContext: u.Tree =?> Boolean = {
+        case Core.Lang.ValDef(sym, _, _) if highFuns contains sym => true
+      }
```
I suggest:

1. Define `isHighContext` as `api.TopDown.inherit(...)` and reuse below.
2. Chain the rest of the transformations with `andThen`.

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+      val ctorsChanged = api.TopDown
+        .inherit(isHighContext)(Monoids.disj)
+        .transformWith {
+        case Attr.inh(api.DefCall(Some(api.ModuleRef(API.bagModuleSymbol)), method, targs, argss), false :: _)
```
Varargs in the pattern match `argss@_*`. Why not use `core.DefCall`?

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+          val targetMethodDecl = backendSymbol.info.decl(method.name)
+          assert(targetMethodDecl.alternatives.size == 1, s"Target method `$targetMethodDecl` shouldn't be overloaded.")
+          val targetMethod = targetMethodDecl.asMethod
+          api.DefCall(Some(api.ModuleRef(backendSymbol)))(targetMethod, targs: _*)(argss)
```
Varargs again `argss: _*`

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+        .inherit(isHighContext)(Monoids.disj)
+        .accumulateWith[Set[u.TermSymbol]] {
+          case Attr.inh(Core.Lang.ValRef(sym), true :: _)
+          if api.Type.of(sym).typeConstructor == API.DataBag =>
```
Use `=:=` to compare types.

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+          if valsRefdFromHigh(lhs)
+          => {
+            val origSym = api.TermSym.free(api.TermName.fresh("orig"), api.Type.of(lhs)) // will be the original bag
+            val ScalaTraversableSym = u.rootMirror.staticModule(s"$rootPkg.api.ScalaTraversable")
```
I suggest extracting the module symbol out of `translateDataflows`.

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+            val origSym = api.TermSym.free(api.TermName.fresh("orig"), api.Type.of(lhs)) // will be the original bag
+            val ScalaTraversableSym = u.rootMirror.staticModule(s"$rootPkg.api.ScalaTraversable")
+            val fetchCall = Core.Lang.DefCall(Some(api.ModuleRef(ScalaTraversableSym))) (
+              ScalaTraversableSym.info.decl(api.TermName("byFetch")).asInstanceOf[u.TermSymbol],
```
`byFetch` can be extracted too.

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+              ScalaTraversableSym.info.decl(api.TermName("byFetch")).asInstanceOf[u.TermSymbol],
+              Core.bagElemTpe(Core.Lang.ValRef(lhs))) (
+              Seq(Core.Lang.ValRef(origSym)))
+            val fetchedSym = api.TermSym.free(api.TermName.fresh("fetched"), api.Type.of(fetchCall))
```
No need to use `free`, you know the owner is `lhs` (see the tree below).

joroKr21 commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/backend/TranslateToDataflows.scala:
```diff
@@ -0,0 +1,107 @@
+          case Attr.inh(api.ValDef(lhs, rhs, flags), false :: _)
+          if valsRefdFromHigh(lhs)
+          => {
+            val origSym = api.TermSym.free(api.TermName.fresh("orig"), api.Type.of(lhs)) // will be the original bag
```
No need to use `free`, you know the owner is `lhs` (see tree below).

joroKr21 commented on emma-spark/src/test/scala/org/emmalanguage/compiler/spark/CodegenTest.scala:
```diff
@@ -0,0 +1,40 @@
+
+  import compiler._
+
+  override def backendPipeline: (compiler.u.Tree) => compiler.u.Tree = {
```
This can be a val.

joroKr21 commented on emma-spark/src/test/scala/org/emmalanguage/compiler/spark/CodegenTest.scala:
```diff
@@ -0,0 +1,40 @@
+    addContext()
+  }
+
+  def addContext(): u.Tree => u.Tree = tree => {
```
This can be a val

joroKr21 commented on emma-spark/src/test/scala/org/emmalanguage/compiler/spark/CodegenTest.scala:
```diff
@@ -0,0 +1,40 @@
+  def addContext(): u.Tree => u.Tree = tree => {
+    import u._
+    q"""
+        implicit val ctx = org.emmalanguage.LocalSparkSession.getOrCreate()
```
Use `_root_` prefix in quasiquotes.

joroKr21 commented on emma-spark/src/test/scala/org/emmalanguage/compiler/spark/CodegenTest.scala:
```diff
@@ -0,0 +1,40 @@
+  override def backendPipeline: (compiler.u.Tree) => compiler.u.Tree = {
+    Comprehension.desugar(API.bagSymbol)
+  } andThen {
+    compiler.Backend.translateToDataflows(universe.rootMirror.staticModule(s"$rootPkg.api.SparkRDD"))
```
Module symbol can be extracted in a val.

ggevay commented on emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala:
```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
+    scala.util.hashing.MurmurHash3.unorderedHash(fetch())
+
+  override def toString: String =
+    fetch().toString
```
I would say yes. This is also how `toString` works on normal Scala collections. (i mean printing the whole collection)

aalexandrov replied:
This a user-facing API, we should be very careful what we expose. For example, if you evaluate a `DataBag` in the Scala REPL, per default the result is rendered using a `toString()` call. 

I suggest to implement this by delegating to `rep.toString`.

ggevay replied:
My thinking here was that we would like to hide it from the user as much as possible that these are distributed collections. Therefore, we probably don't want different semantics for the different backends (e.g. something like `DataSet@45754` in case of Flink).

But I just had a new idea: Some REPL-style environments (e.g. Maple) use the following strategy:
- if the result is small then they fully print it
- otherwise they print only a small part of it, and then print `<and 234627635 more elements>`

Maybe we could also do it like this.

ggevay commented on emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala:
```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
```
At the moment, I think we would just use this probably for debugging, on small bags, so it's not really important to optimize it. But I'll add a TODO comment to make this smarter.

What we should actually do here in the long term is to write this without fetching at all. But this would complicate all the generated jobs that use this, and I don't want to do that at the moment.

ggevay replied:
As with the `hashCode`, we could also provide a non-abstract, yet scalable implementation of this here: `(this minus that).isEmpty && (that minus this).isEmpty`.

ggevay replied:
Hm, it seems that we don't have `minus` in the new API for some reason. Should I add it?

aalexandrov replied:
I removed it deliberately for a number of reasons:

1. None of the backend API have native support for difference (and we didn't support translating `minus` in the old backend).
- It was never used in any of the examples.

aalexandrov replied:
Implementing full equality check (case 2 of my suggested `equals` structure) can be tricky. As this will almost never occur, we can still compare after fetching locally for now, but we should open an issue for that and resolve this in the next weeks.

ggevay replied:
> None of the backend API have native support for difference (and we didn't support translating minus in the old backend).

We could do it with a `coGroup`.

> It was never used in any of the examples.

Well, we would use it here for implementing the equality check :)

aalexandrov replied:
But from the three backends only Spark's Dataset has native support for that, for the others, you'll have to implement it with `coGroup` similar to what I suggested above.

Also

```
(this minus that).isEmpty && (that minus this).isEmpty
```

will require two co-groups, whereas the variant I sketched can work with one.

My other concern is that minus makes reasoning about logical rewrites a bit harder, and it is not needed for a lot of use cases we can make our life simpler by simply not including it (at least until we see a strong need for the opposite).

ggevay replied:
> we can make our life simpler by simply not including it

I'm not sure about this. The problem is that the user can actually use `exists` to implement `minus`, so I would say that if our rewrites can't handle these situations, that is a problem, regardless of whether `minus` is actually in the API.

aalexandrov replied:
> The problem is that the user can actually use exists to implement minus

A definition such as

```scala
def minus(that: DataBag[A]): DataBag[A] = for {
  x <- xs; if ys.exists(x)
} yield x
```

opens the door for inconsistencies. For example with

```scala
def equals(that: DataBag[A]): Boolean =
  this.size == that.size && (this minus that).isEmpty && (that minus this).isEmpty
```

you will get wrong results.

```scala
DataBag(Seq(1,2,2,3)) equals DataBag(Seq(1,1,2,3)) 
```

To make `minus` compatible with the `equals` definition, it should be modified to account for the multiplicities. 

The more general problem is that defining a Boolean algebra on bags is not consistent with the usual semantics of `union` taken by the targeted backends, which is more efficient to implement and sufficient for most real-world scenarios.

ggevay replied:
Sorry, I missed the multiplicities. But still, it's possible to implement `minus`, just a little more involved:

First, we convert both bags to the representation `(value, multiplicity)`, e.g. the bag
```
a
a
b
```
becomes
```
(a,2)
(b,1)
```
This is possible with a `groupBy` and then `map size`.
Then we can use an inner join to check that the multiplicities of those elements that appear in both bags match, and then use exists to check for values that appear in only one of them.

ggevay commented on emma-language/src/main/scala/org/emmalanguage/compiler/lang/core/Core.scala:
```diff
@@ -367,8 +367,10 @@ trait Core extends Common
     def bagElemTpe(xs: u.Tree): u.Type = {
-      assert(xs.tpe.typeConstructor == API.DataBag)
-      api.Type.arg(1, xs.tpe)
+      assert(xs.tpe.typeConstructor == API.DataBag,
```
Yes, sorry, I'll fix this.

ggevay commented:
Thanks for the review @joroKr21! I think I've addressed all the comments (eb8f472).
Additionally, I moved the two `CodegenTest` files into better named packages.

aalexandrov commented on emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala:
```diff
@@ -313,6 +318,36 @@ trait DataBag[A] extends Serializable {
+  }
+
+  // TODO: Do this with a fold.
+  override def hashCode(): Int =
```
Similar to `equals`, this should be kept abstract here and implemented at each `DataBag` implementation. 

For non-local `DataBag` variants, one can use a flavor of [this snippet](https://gist.github.com/aalexandrov/089f0fd4d3de13d37536f035dfe14c19).

aalexandrov replied:
Agreed.

aalexandrov commented on emma-flink/src/main/scala/org/emmalanguage/api/FlinkDataSet.scala:
```diff
@@ -101,7 +101,7 @@ class FlinkDataSet[A: Meta] private[api](@transient private val rep: DataSet[A])
   }
 
-  def fetch(): Seq[A] =
+  protected override def collect(): Traversable[A] =
```
Why not use

```scala
lazy val collect: Traversable[A] = 
  rep.collect()
```

Then you can get rid of the `collected` member as well?

aalexandrov commented on emma-language/src/main/scala/org/emmalanguage/api/ScalaTraversable.scala:
```diff
@@ -80,27 +80,8 @@ class ScalaTraversable[A] private[api](private val rep: Traversable[A]) extends
-  override def fetch(): Traversable[A] =
+  protected override def collect(): Traversable[A] =
     rep
-
```
Revert this delete.

aalexandrov commented on emma-language/src/main/scala/org/emmalanguage/compiler/Compiler.scala:
```diff
@@ -65,6 +65,24 @@ trait Compiler extends AlphaEq with Source with Core with Backend {
-    Function.chain(bld.result())
+    val steps = bld.result()
+
+    if (!printAllTrees) {
```
No curly brackets needed here

```scala
if (!printAllTrees) Function.chain(steps)
else Function.chain(List(print) ++ steps.flatMap(List(_, print)))
```

ggevay replied:
I would agree with you, but unfortunately Scalastyle disagrees... (unless I write the whole `if else` thing on one line, but that hurts the readability).

ggevay replied:
Should I just try to disable this inspection as well?

aalexandrov commented:
A squash commit that implements hashing as proposed [can be found here](https://github.com/aalexandrov/emma/commits/base-codegen-test).

ggevay commented:
OK, I think I have addressed all comments. The problematic points:
- `toString`: We discussed offline with @aalexandrov, and came to the conclusion that we don't really want to hide it from the user that `DataBag` is a lazy collection. Therefore, for now, `toString` will print whatever the implementation does, which is not the contents for the Flink ans Spark implementations. A concrete example where this behaviour works better is a Scala REPL.
- `equals`: I'm doing `fetch` in the trait for now. Later, we can think about having
  - a scalable (non-fetch) implementation of it in the trait based on the API that is in the trait
  - and/or efficient implementations in the backends based on `coGroup`
- `hashCode`: I added a default implementation that does a `fetch` and then `MurmurHash3.unorderedHash`. I also added the backend-specific overrides of @aalexandrov that use `mapPartition` to make this faster. (I think also having the default implementation is useful as a specification, i.e. for specifying that the backend implementations should always give the same result as `MurmurHash3.unorderedHash` would give on the fetched collection.

There is also the question of whether we want the different implementations to be able to interoperate, e.g. can a `FlinkDataSet` equal a `SparkRDD`. At the moment it can, but this can be up to future debate. (Note, that at the moment this is inconsistent with how e.g. `union` work, because `union` fails if called on different backend implementations.) A consideration about this is whether we want to allow `DataBag` to escape from a parallelize block. If yes, then it would probably make sense to make it fool-proof by hiding implementation-specific behaviour like issues with equality between a `FlinkDataSet` and a `SparkRDD`.

ggevay commented:
The tests were failing because of a silly bug, which should be fixed now.

aalexandrov commented:
LGTM. I will follow up with a PR that addresses `equals`, value semantics and `DataBag` subclass compatibility for `union`.

ggevay commented:
OK, thanks!

aalexandrov commented:
@joroKr21 can you approve your review? I think most of the comments were addressed.

ggevay commented:
I rebased to newir, and several tests are failing in `BaseCodegenTest`. There have only been two commits introduced by the rebase, and the relevant one is probably this abd8ca4e30ce13109b12e18cba573e10b19f3b64.

Some error msgs that I'm getting:
```

Error at NoPosition: Cannot convert type `(Int, Int) @unchecked` to a type-tree
java.lang.RuntimeException: Error at NoPosition: Cannot convert type `(Int, Int) @unchecked` to a type-tree
	at org.emmalanguage.ast.JavaAST$class.abort(JavaAST.scala:47)
	at org.emmalanguage.compiler.RuntimeCompiler.abort(RuntimeCompiler.scala:27)
	at org.emmalanguage.compiler.RuntimeCompiler.abort(RuntimeCompiler.scala:27)
	at org.emmalanguage.ast.Types$TypeAPI$Type$.original$1(Types.scala:395)
	at org.emmalanguage.ast.Types$TypeAPI$Type$.tree(Types.scala:403)
	at org.emmalanguage.ast.AST$$anonfun$restoreTypeTrees$1.applyOrElse(AST.scala:90)
	at org.emmalanguage.ast.AST$$anonfun$restoreTypeTrees$1.applyOrElse(AST.scala:89)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at org.emmalanguage.ast.Transversers$Strategy$$anonfun$forgetful$1.applyOrElse(Transversers.scala:343)
	at org.emmalanguage.ast.Transversers$Strategy$$anonfun$forgetful$1.applyOrElse(Transversers.scala:342)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:34)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1400)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.internal.Trees$$anonfun$itransform$1.apply(Trees.scala:1357)
	at scala.reflect.internal.Trees$$anonfun$itransform$1.apply(Trees.scala:1356)
	at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1355)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.api.Trees$Transformer$$anonfun$transformStats$1.apply(Trees.scala:2589)
	at scala.reflect.api.Trees$Transformer$$anonfun$transformStats$1.apply(Trees.scala:2587)
	at scala.collection.immutable.List.loop$1(List.scala:173)
	at scala.collection.immutable.List.mapConserve(List.scala:189)
	at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1366)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.internal.Trees$$anonfun$itransform$1.apply(Trees.scala:1357)
	at scala.reflect.internal.Trees$$anonfun$itransform$1.apply(Trees.scala:1356)
	at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1355)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.api.Trees$Transformer$$anonfun$transformStats$1.apply(Trees.scala:2589)
	at scala.reflect.api.Trees$Transformer$$anonfun$transformStats$1.apply(Trees.scala:2587)
	at scala.collection.immutable.List.loop$1(List.scala:173)
	at scala.collection.immutable.List.mapConserve(List.scala:189)
	at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1366)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.internal.Trees$$anonfun$itransform$3.apply(Trees.scala:1379)
	at scala.reflect.internal.Trees$$anonfun$itransform$3.apply(Trees.scala:1379)
	at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1378)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.api.Trees$Transformer$$anonfun$transformTrees$1.apply(Trees.scala:2559)
	at scala.reflect.api.Trees$Transformer$$anonfun$transformTrees$1.apply(Trees.scala:2559)
	at scala.collection.immutable.List.loop$1(List.scala:173)
	at scala.collection.immutable.List.mapConserve(List.scala:189)
	at scala.reflect.api.Trees$Transformer.transformTrees(Trees.scala:2559)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1347)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at scala.reflect.internal.Trees$class.itransform(Trees.scala:1366)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at org.emmalanguage.ast.Transversers$Transform.org$emmalanguage$ast$Transversers$Transform$$super$transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$$anonfun$transform$1.apply(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$ManagedAttr$class.at(Transversers.scala:155)
	at org.emmalanguage.ast.Transversers$Transform.at(Transversers.scala:348)
	at org.emmalanguage.ast.Transversers$Transform.transform(Transversers.scala:359)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.org$emmalanguage$ast$Transversers$Transform$topDown$$anon$$super$transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3$$anonfun$transform$2.apply(Transversers.scala:554)
	at org.emmalanguage.util.Functions$$anonfun$compose$1.applyOrElse(Functions.scala:33)
	at org.emmalanguage.ast.Transversers$Transform$topDown$$anon$3.transform(Transversers.scala:554)
	at org.emmalanguage.ast.Transversers$Transform.apply(Transversers.scala:355)
	at org.emmalanguage.ast.Transversers$Transform.apply(Transversers.scala:348)
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
	at scala.Function$$anonfun$chain$1$$anonfun$apply$1.apply(Function.scala:24)
	at scala.Function$$anonfun$chain$1$$anonfun$apply$1.apply(Function.scala:24)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:151)
	at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
	at scala.Function$$anonfun$chain$1.apply(Function.scala:24)
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
	at org.emmalanguage.compiler.BaseCodegenTest.verify(BaseCodegenTest.scala:57)
	at org.emmalanguage.compiler.BaseCodegenTest$$anonfun$11$$anonfun$apply$mcV$sp$22.apply$mcV$sp(BaseCodegenTest.scala:311)
	at org.emmalanguage.compiler.BaseCodegenTest$$anonfun$11$$anonfun$apply$mcV$sp$22.apply(BaseCodegenTest.scala:311)
	at org.emmalanguage.compiler.BaseCodegenTest$$anonfun$11$$anonfun$apply$mcV$sp$22.apply(BaseCodegenTest.scala:311)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FreeSpecLike$$anon$1.apply(FreeSpecLike.scala:373)
	at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
	at org.scalatest.FreeSpec.withFixture(FreeSpec.scala:1740)
	at org.scalatest.FreeSpecLike$class.invokeWithFixture$1(FreeSpecLike.scala:370)
	at org.scalatest.FreeSpecLike$$anonfun$runTest$1.apply(FreeSpecLike.scala:382)
	at org.scalatest.FreeSpecLike$$anonfun$runTest$1.apply(FreeSpecLike.scala:382)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.FreeSpecLike$class.runTest(FreeSpecLike.scala:382)
	at org.emmalanguage.compiler.BaseCodegenTest.org$scalatest$BeforeAndAfter$$super$runTest(BaseCodegenTest.scala:37)
	at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:200)
	at org.emmalanguage.compiler.BaseCodegenTest.runTest(BaseCodegenTest.scala:37)
	at org.scalatest.FreeSpecLike$$anonfun$runTests$1.apply(FreeSpecLike.scala:441)
	at org.scalatest.FreeSpecLike$$anonfun$runTests$1.apply(FreeSpecLike.scala:441)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:390)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:427)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
	at org.scalatest.FreeSpecLike$class.runTests(FreeSpecLike.scala:441)
	at org.scalatest.FreeSpec.runTests(FreeSpec.scala:1740)
	at org.scalatest.Suite$class.run(Suite.scala:1424)
	at org.scalatest.FreeSpec.org$scalatest$FreeSpecLike$$super$run(FreeSpec.scala:1740)
	at org.scalatest.FreeSpecLike$$anonfun$run$1.apply(FreeSpecLike.scala:486)
	at org.scalatest.FreeSpecLike$$anonfun$run$1.apply(FreeSpecLike.scala:486)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
	at org.scalatest.FreeSpecLike$class.run(FreeSpecLike.scala:486)
	at org.emmalanguage.compiler.BaseCodegenTest.org$scalatest$BeforeAndAfter$$super$run(BaseCodegenTest.scala:37)
	at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:241)
	at org.emmalanguage.compiler.BaseCodegenTest.run(BaseCodegenTest.scala:37)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
	at org.scalatest.tools.Runner$.run(Runner.scala:883)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:138)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

Another one:
```
scala.tools.reflect.ToolBoxError: Compilation failed for tree:
================
class `8f75bdfa-2791-48ae-9165-0b22c30e7e23` {
  def run() = {
    val eu = "eu";
    val com = "com";
    val java = "java";
    val org = "org";
    val scala = "scala";
    _root_.org.emmalanguage.api.DataBag.apply[scala.Int](_root_.scala.Predef.intWrapper(0).to(100)).sum(_root_.scala.math.Numeric.IntIsIntegral)
  }
}
================

	at org.emmalanguage.ast.JavaAST$class.compile(JavaAST.scala:84)
	at org.emmalanguage.compiler.RuntimeCompiler.compile(RuntimeCompiler.scala:27)
	at org.emmalanguage.compiler.BaseCompilerSpec$$anonfun$1.apply(BaseCompilerSpec.scala:47)
	at org.emmalanguage.compiler.BaseCompilerSpec$$anonfun$1.apply(BaseCompilerSpec.scala:43)
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
	at org.emmalanguage.compiler.BaseCodegenTest.verify(BaseCodegenTest.scala:57)
	at org.emmalanguage.compiler.BaseCodegenTest$$anonfun$15$$anonfun$apply$mcV$sp$39.apply$mcV$sp(BaseCodegenTest.scala:522)
	at org.emmalanguage.compiler.BaseCodegenTest$$anonfun$15$$anonfun$apply$mcV$sp$39.apply(BaseCodegenTest.scala:522)
	at org.emmalanguage.compiler.BaseCodegenTest$$anonfun$15$$anonfun$apply$mcV$sp$39.apply(BaseCodegenTest.scala:522)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FreeSpecLike$$anon$1.apply(FreeSpecLike.scala:373)
	at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
	at org.scalatest.FreeSpec.withFixture(FreeSpec.scala:1740)
	at org.scalatest.FreeSpecLike$class.invokeWithFixture$1(FreeSpecLike.scala:370)
	at org.scalatest.FreeSpecLike$$anonfun$runTest$1.apply(FreeSpecLike.scala:382)
	at org.scalatest.FreeSpecLike$$anonfun$runTest$1.apply(FreeSpecLike.scala:382)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.FreeSpecLike$class.runTest(FreeSpecLike.scala:382)
	at org.emmalanguage.compiler.BaseCodegenTest.org$scalatest$BeforeAndAfter$$super$runTest(BaseCodegenTest.scala:37)
	at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:200)
	at org.emmalanguage.compiler.BaseCodegenTest.runTest(BaseCodegenTest.scala:37)
	at org.scalatest.FreeSpecLike$$anonfun$runTests$1.apply(FreeSpecLike.scala:441)
	at org.scalatest.FreeSpecLike$$anonfun$runTests$1.apply(FreeSpecLike.scala:441)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:390)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:427)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
	at org.scalatest.FreeSpecLike$class.runTests(FreeSpecLike.scala:441)
	at org.scalatest.FreeSpec.runTests(FreeSpec.scala:1740)
	at org.scalatest.Suite$class.run(Suite.scala:1424)
	at org.scalatest.FreeSpec.org$scalatest$FreeSpecLike$$super$run(FreeSpec.scala:1740)
	at org.scalatest.FreeSpecLike$$anonfun$run$1.apply(FreeSpecLike.scala:486)
	at org.scalatest.FreeSpecLike$$anonfun$run$1.apply(FreeSpecLike.scala:486)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
	at org.scalatest.FreeSpecLike$class.run(FreeSpecLike.scala:486)
	at org.emmalanguage.compiler.BaseCodegenTest.org$scalatest$BeforeAndAfter$$super$run(BaseCodegenTest.scala:37)
	at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:241)
	at org.emmalanguage.compiler.BaseCodegenTest.run(BaseCodegenTest.scala:37)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
	at org.scalatest.tools.Runner$.run(Runner.scala:883)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:138)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: scala.tools.reflect.ToolBoxError: reflective compilation has failed:

type Int is not a member of String
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.throwIfErrors(ToolBoxFactory.scala:316)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.wrapInPackageAndCompile(ToolBoxFactory.scala:198)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.compile(ToolBoxFactory.scala:252)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$$anonfun$compile$2.apply(ToolBoxFactory.scala:429)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$$anonfun$compile$2.apply(ToolBoxFactory.scala:422)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$.liftedTree2$1(ToolBoxFactory.scala:355)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$withCompilerApi$.apply(ToolBoxFactory.scala:355)
	at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl.compile(ToolBoxFactory.scala:422)
	at org.emmalanguage.ast.JavaAST$class.compile(JavaAST.scala:80)
	... 66 more
```

joroKr21 commented:
The first error is due to a missing case in type -> type-tree conversion (`AnnotatedType`). I will add this and it should be fixed. The second problem is a missing `_root_`. To solve this change [this line](https://github.com/emmalanguage/emma/blob/newir/emma-language/src/test/scala/org/emmalanguage/compiler/BaseCompilerSpec.scala#L46) to either `api.Tree.show(wrapped)` or `u.showCode(wrapped, printRootPkg = true)`.

There is a third problem. Somehow the type `Movies.this.ImdbYear` appears, even though `ImdbYear` is statically accessible.

aalexandrov commented:
I am working on the `AnnotatedType` case.

aalexandrov commented:
I managed to fix the `AnnotatedType` issue related-test ([see here](https://github.com/aalexandrov/emma/commit/e42fba9dd57d58894ec43dc927938b22666ac549#diff-a1e5f12e70b140f3e273186106e7945fR397)), but the proper solution (`res1`) gives me another error. Will continue after lunch.

joroKr21 commented:
Solution for the weird this types - the first case should change to this:
```scala
// Degenerate type: `this[staticID]`.
            case u.ThisType(sym) if sym.isStatic =>
              api.Tree.resolveStatic(sym)
```

joroKr21 commented:
There is a bug in [PatternMatching](https://github.com/emmalanguage/emma/blob/newir/emma-language/src/main/scala/org/emmalanguage/compiler/lang/source/PatternMatching.scala#L83). `target` should be replaced by `unascr`.

Now that we solved the problems who should commit the solutions and to which branch?

aalexandrov commented:
@joroKr21 I'm on it, but the `AnnotatedType` case will need a more general fix.

joroKr21 commented:
What's wrong with the `foldLeft` over the annotations?

aalexandrov commented:
It produces a weird tree such as:

```
val x = (Tuple2(Int,Int) @unchecked)
```

I am now filtering all `unchecked` annotations such that for the problematic cases in @ggevay's case the fold actually does not append anything.

```scala
// Annotated type: `A @ann1 @ann2 ...`
case AnnotatedType(annotations, underlying) =>
  annotations.foldLeft(Type.tree(underlying))((res, ann) =>
    if (ann.tree.tpe =:= typeOf[unchecked]) res
    else u.Annotated(ann.tree, res) // FIXME: don't bypass `@unchecked`
  )
```

This should make the PR good for merging, but the `FIXME` needs deeper investigation. I think it might be related to some other bug further down the pipeline (e.g. in yet another peculiar thing in `showCode`).

joroKr21 commented:
The bug is in `PatternMatching` (see my comment above). This `@unchecked` shouldn't have been there in the first place. It seems that `showCode` is treating `@unchecked` specially, because the tree looks differently if you debug it.

Other than that, you should use `original(underlying)` to seed the fold, because `Type.tree` wraps it in `TypeTree().setOriginal`.

aalexandrov commented:
@joroKr21 you're right, the `PatternMatching` fix seems to resolve the issue.
